mgnify_pipelines_toolkit/analysis/assembly/mgnify2mimc_inputs.py

```diff
@@ -0,0 +1,726 @@
+                    data["id"]
+                    for data in api_endpoint["data"]
+                    if data["attributes"]["experiment-type"] == "assembly"
+                ]
+
+                break
+            else:
+                pass
+                # print(f"Attempt {attempt + 1} failed with status code: {response.status_code}")
+        except requests.exceptions.RequestException as e:
+            print(f"Attempt {attempt + 1} failed due to error: {e}")
+
+        if attempt < max_retries - 1:
+            print(f"Retrying in {retry_delay} seconds...")
+            time.sleep(retry_delay)
+
+        else:
+            print(
+                "All retries failed. Please check the MGYS study accesion or try again later."
+            )
+
+    return mgya_ids
+
+
+# download pfam results for a given MGYA accession
+def download_pfam_from_mgya(mgya_id):
+
+    max_retries = 3  # max retires for url accessing
+    # retry_delay = 5
+
+    assembly_id = " "
+    sample_id = " "
+
+    # api endpoint for downloads
+    url = f"https://www.ebi.ac.uk/metagenomics/api/v1/analyses/{mgya_id}/downloads"
+
+    mgya_pfam_df = pd.DataFrame(columns=["protein", "PfamID", "pfam_description"])
+
+    mgya_pfams = []
+
+    for attempt in range(max_retries):
+        try:
+            response = requests.get(url)
+            if response.status_code == 200:
+                for downloadable_file in response.json()["data"]:
+
+                    if (
+                        "pfam"
+                        in downloadable_file["attributes"]["description"][
+                            "label"
+                        ].lower()
+                    ):
+
+                        # url of pfam file
+                        url = downloadable_file["links"]["self"]
+                        mgya_pfam_df = pd.read_csv(
+                            url,
+                            sep=",",
+                            names=["protein", "PfamID", "pfam_description"],
+                        )
+
+                        mgya_pfams = list(mgya_pfam_df["PfamID"])
+
+                        break
+            else:
+                print(
+                    f"Connection attempt failed with status code: {response.status_code}"
+                )
+        except requests.exceptions.RequestException as e:
+            print(f"Connection attempt {attempt + 1} failed due to error: {e}")
+
+        try:
+            response = requests.get(
+                f"https://www.ebi.ac.uk/metagenomics/api/v1/analyses/{mgya_id}"
+            )
+            if response.status_code == 200:
+                mgya_metadata = response.json()["data"]["relationships"]
+                sample_id = mgya_metadata["sample"]["data"]["id"]
+                assembly_id = mgya_metadata["assembly"]["data"]["id"]
+            else:
+                print(
+                    f"Attempt {attempt + 1} failed with status code: {response.status_code}"
+                )
```

Comment:
I think you do not need that separate request here. You have already requested `https://www.ebi.ac.uk/metagenomics/api/v1/analyses/{mgya_id}/` as `response.json()["data"]`. You can get downloads and relationships from there and process in one request